### ISSUE TITLE
Evaluation when using pre trained vectors

### ISSUE THREAD

**Issue by naveenjafer** [dennybritz/cnn-text-classification-tf#116]
Hi Denny, 
            I have been using word2vec pre trained vectors with multi class classification, I have one query that has still not been solved that is related to the eval.py.
Say one of my test case has a set of words that are synonyms to one of the categories in the training case, then in that case the classifier will not be able to produce the desired classification/result right?
Since those words do not exist in the dictionary, the row in x_test corresponding to that use case will be a vector of 0's. But shouldn't it mapping to the word2vec dictionary and picking out its weight vector and using that while processing? 

Basically the point I am trying to make is, say there is a word "motivation" in the training set. And a new word "incentive" comes in my testing set that the vocabulary has not seen before, then this word is not treated any different from any other unseen word like "humour". Bot "humour" and "incentive" were not present in the vocab and they end up being allotted random numbers weights, right? Incentive is a very close synonym of motivation and might very well have had a similar embedding in the word2vec list.

**Comment by naveenjafer** [comment]
The modified code that I have used for the same is https://github.com/naveenjafer/cnn-text-classification-tf

**Comment by naveenjafer** [comment]
@yoonkim Would you be able to help with this?

**Comment by jagganaidu** [comment]
Hi Naveenjafer:

Do you have a solution for this? I am running into the same issue. If I encounter new words in the Evaluation Data, the model is not able to figure it out though the word has pre-trained representation in GloVe. 
Thanks for your help.
Jagan

**Comment by naveenjafer** [comment]
Hi Jagan, Yes I did manage to get a crude workaround to it, The problem lies in the fact that the new word doesnt exist in the vocab_processor that is used to get the word embeddings for a word not in the training set, do the following 2 things.
1) Create the vocab_processor with some reasonable max length of your choice( I used 40) 
`vocab_processor2 = learn.preprocessing.VocabularyProcessor(max_document_length)`
Now, when you fit this use not just your training set, but create a list of sentences making use of the words in GoogleWord2Vec or any embedding you use, 
`vocab_processor2.fit_transform(x_text_vocab + x_text)` -> Here, x_text is the list of sentences that I actually want to train my model with and x_text_vocab would just be a list of sentences that neednt make any sense whatsoever, it could be a list with one word from the GoogleWord2Vec. 

2) Now, you will have to construct the 'x' array that is used for training on your own,
`for i,item in enumerate(x_text):
    sentence_list = item.split(" ")
    for j,word in enumerate(sentence_list):
        print(vocab_dict[word])
        x[i][j] = vocab_dict[word]
 print(x)`
Now x is the 2D array with the embedding for each word of the sentence. One training sentence in the row and all columns are stacked up to represent the entire training set.

Hope this helps. :)

**Comment by jagganaidu** [comment]
Thanks Naveen! I presume you go by that name! It makes sense. 

I was thinking slightly a different route. In the evaluation, we can have an if statement to see if it's an unknown word. If that's the case, check in GoogleWord2Vec for the corresponding embedding (presuming it appears there) and bring the corresponding embedding and do the evaluation. However, not sure how we can take advantage of the VocabularyProcessor capabilities in there. Not sure I am clear enough. However, if it makes sense please let me know whether there is a way forward. My expertise in this area is rather limited :)

**Comment by naveenjafer** [comment]
Yes, that is a possibility that would work too, but since I had spent more time on the code for the train.py for modifying it for my use case of multi classes, I went ahead and did that modification in the initial parts of training, and also to reduce any non optimized overheads from happening during evaluation for speed.

**Comment by jagganaidu** [comment]
Hi Naveen:
Thanks for making the point of the evaluation speed, which would be affected if I have to change the code in the eval.py. 

In your approach, let me go back to your example of incentive and motivation. If "motivation" is in the training set and incentive is something that you added (but not trained I presume) in the random additional sentences that you have, wouldn't the cosine distance between these two words change after training because "motivation" is trained and "incentive" is not. It may not make much of a difference but thought of running it through you.
On a slightly different note, do you think it's worthwhile to pay attention to spelling? If so, how do you deal with it? The example that I am working on has a lot of spelling issues. Please let me know.

**Comment by naveenjafer** [comment]
Hi Jagan, So in this implementation followed by denny, this is a static training where the initial word 2 vecs presented do not change with the training, so the cosine distance between a word like motivation and incentive doesn't change through the course of the training, but if you go back to the original paper by Yoon Kim, he also gives an architecture option of non static vectors where the word 2 vector is also retrained through the course of model creation, the point that you raised would occur in that case. It is quite important to pay attention to spelling too, and I have plans to implement a separate spelling model for the same that would operate on the input sentence before putting it through the model. I haven't been able to spend much time on it though.
